dvc/commands/data.py

```diff
@@ -0,0 +1,387 @@
+    cache: "HashFileDB",
+) -> Dict[str, List[str]]:
+    # TODO: add support for shallow diff in dvc-data
+    # TODO: we may want to recursively do in_cache check
+    d = {}
+
+    from dvc_data.objects.tree import Tree
+
+    if isinstance(new_obj, Tree):
+        root = os.path.sep.join([root, ""])
+
+    if not _in_cache(old_obj, cache):
+        d["not_in_cache"] = [root]
+
+    if old_obj is None and new_obj is None:
+        return d
+    if old_obj is None:
+        return {"added": [root], **d}
+    if new_obj is None:
+        return {"deleted": [root], **d}
+    if old_obj.hash_info != new_obj.hash_info:
+        return {"modified": [root], **d}
+    return {"unchanged": [root], **d}
+
+
+def granular_diff(
+    root: str,
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+    with_dirs: bool = False,
+) -> Dict[str, List[str]]:
+    from dvc_data.diff import ROOT
+    from dvc_data.diff import diff as odiff
+    from dvc_data.objects.tree import Tree
+
+    def path_join(root: str, *paths: str) -> str:
+        if not isinstance(new_obj, Tree):
+            return root
+        return os.path.sep.join([root, *paths])
+
+    diff_data = odiff(old_obj, new_obj, cache)
+    drop_root = not with_dirs and isinstance(new_obj, Tree)
+
+    output: Dict[str, List[str]] = defaultdict(list)
+    for state in ("added", "deleted", "modified", "unchanged"):
+        items = getattr(diff_data, state)
+        output[state].extend(
+            path_join(root, *item.new.key)
+            for item in items
+            if not (drop_root and item.new.key == ROOT)
+        )
+        # TODO: PERF: diff is checking not_in_cache for each even if we only
+        # need it for the index.
+        # BUG: not_in_cache file also shows up as modified in staged and
+        # unstaged. We currently don't know if it is really modified.
+        output["not_in_cache"].extend(
+            path_join(root, *item.new.key)
+            for item in items
+            if not item.old.in_cache
+            and not (drop_root and item.new.key == ROOT)
+            and state != "added"
+        )
+    return output
+
+
+def diff(
+    root: str,
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+    granular: bool = False,
+    with_dirs: bool = False,
+) -> Dict[str, List[str]]:
+    if granular:
+        return granular_diff(
+            root, old_obj, new_obj, cache, with_dirs=with_dirs
+        )
+    return shallow_diff(root, old_obj, new_obj, cache)
+
+
+class CmdDataStatus(CmdBase):
+    COLORS = {
+        "not_in_cache": "red",
+        "unstaged": "yellow",
+        "staged": "green",
+        "untracked": "cyan",
+    }
+    LABELS = {
+        "not_in_cache": "Not in cache",
+        "staged": "DVC committed changes",
+        "unstaged": "DVC uncommitted changes",
+        "untracked": "Untracked files",
+        "unchanged": "DVC unchanged files",
```

Comment:
Could be more descriptive. Ideas:
```suggestion
        "not_in_cache": "Not in DVC cache",
        "staged": "Committed changes to files tracked by DVC",
        "unstaged": "Uncommitted changes to files tracked by DVC",
        "untracked": "Files not tracked by DVC (or Git)",
        "unchanged": "Existing files tracked by DVC (no changes)",
```

Reply:
It's a bit verbose, to be honest. I think we can simplify it to:
```suggestion
        "not_in_cache": "Not in cache",
        "staged": "Changes tracked by DVC",
        "unstaged": "Changes not tracked by DVC",
```

I am not sure about `untracked` and `unchanged`. `not_in_cache` could stay simpler (I don't think `dvc` is adding any information, `cache` implies `dvc cache`).